File: problems/图/133.克隆图.py
#!/usr/bin/env python3
"""
133. 克隆图
"""
class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors else []

def clone_graph(node):
    if not node:
        return None
    visited = {}
    def dfs(n):
        if n in visited:
            return visited[n]
        clone = Node(n.val)
        visited[n] = clone
        for neighbor in n.neighbors:
            clone.neighbors.append(dfs(neighbor))
        return clone
    return dfs(node)

if __name__ == "__main__":
    # 测试用例
    # 构建节点 1->2->4, 1->3->4
    node1 = Node(1)
    node2 = Node(2)
    node3 = Node(3)
    node4 = Node(4)
    node1.neighbors = [node2, node3]
    node2.neighbors = [node4]
    node3.neighbors = [node4]
    cloned = clone_graph(node1)
    print(cloned.val)  # 预期: 1
    print(len(cloned.neighbors))  # 预期: 2

File: problems/图/133.克隆图.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() : val(0), neighbors(vector<Node*>()) {}
    Node(int x) : val(x), neighbors(vector<Node*>()) {}
    Node(int x, vector<Node*> n) : val(x), neighbors(n) {}
};

unordered_map<Node*, Node*> mp;

Node* cloneGraph(Node* node) {
    if (!node) return nullptr;
    if (mp.count(node)) return mp[node];
    Node* clone = new Node(node->val);
    mp[node] = clone;
    for (Node* n : node->neighbors)
        clone->neighbors.push_back(cloneGraph(n));
    return clone;
}

int main() {
    // 测试用例
    // 构建: 1--2, 1--3, 2--4, 3--4
    Node* n1 = new Node(1);
    Node* n2 = new Node(2);
    Node* n3 = new Node(3);
    Node* n4 = new Node(4);
    n1->neighbors = {n2, n3};
    n2->neighbors = {n4};
    n3->neighbors = {n4};
    Node* cloned = cloneGraph(n1);
    cout << cloned->val << endl;  // 预期: 1
    cout << cloned->neighbors.size() << endl;  // 预期: 2
    return 0;
}
